File: tests/fixtures/sample.dart
/// Sample Dart file exercising all extractor features.

library sample;

import 'dart:async';
import 'dart:convert';

const int maxRetries = 3;

typedef JsonMap = Map<String, dynamic>;
typedef Callback = void Function(String message);

/// Represents a log level.
enum LogLevel {
  debug,
  info,
  warning,
  error,
}

/// Interface-like abstract class for serializable objects.
abstract class Serializable {
  JsonMap toJson();
  String toJsonString() => jsonEncode(toJson());
}

/// Mixin that adds timestamps to entities.
mixin Timestamped {
  DateTime get createdAt;
  DateTime? get updatedAt;

  Duration age() => DateTime.now().difference(createdAt);
}

/// A user entity with serialization and timestamp support.
class User extends Serializable with Timestamped {
  final String id;
  final String name;
  final String _email;
  LogLevel logLevel;

  @override
  final DateTime createdAt;

  @override
  DateTime? updatedAt;

  User(this.id, this.name, this._email, {this.logLevel = LogLevel.info})
      : createdAt = DateTime.now();

  User.guest()
      : id = '0',
        name = 'Guest',
        _email = 'guest@example.com',
        logLevel = LogLevel.debug,
        createdAt = DateTime.now();

  @override
  JsonMap toJson() => {
        'id': id,
        'name': name,
        'logLevel': logLevel.name,
      };

  /// Fetch the user profile asynchronously.
  Future<JsonMap> fetchProfile() async {
    await Future.delayed(Duration(milliseconds: 100));
    print('Fetched profile for $name');
    return toJson();
  }

  bool get _isValid => name.isNotEmpty && _email.contains('@');

  void _logAction(String action) {
    print('[$logLevel] $name: $action');
  }
}

/// Extension adding utility methods to String.
extension StringUtils on String {
  String toSlug() => toLowerCase().replaceAll(' ', '-');
  bool get isBlank => trim().isEmpty;
}

/// Top-level function that processes users.
Future<List<JsonMap>> processUsers(List<User> users) async {
  final results = <JsonMap>[];
  for (final user in users) {
    final profile = await user.fetchProfile();
    results.add(profile);
  }
  return results;
}

/// Synchronous helper function.
void logMessage(String message, {LogLevel level = LogLevel.info}) {
  print('[${level.name}] $message');
}

File: tests/fixtures/sample.pas
{ Sample Pascal unit exercising all extractor features. }
unit SampleUnit;

interface

uses
  SysUtils, Classes;

const
  MaxRetries = 3;
  AppName = 'SampleApp';

type
  TLogLevel = (llDebug, llInfo, llWarning, llError);

  TPoint = record
    X: Double;
    Y: Double;
  end;

  { Interface for serializable objects. }
  ISerializable = interface
    function ToJSON: string;
  end;

  { Abstract base class for entities. }
  TEntity = class
  private
    FId: string;
    FCreatedAt: TDateTime;
  protected
    function GetId: string;
  public
    constructor Create(const AId: string);
    destructor Destroy; override;
    function Validate: Boolean; virtual; abstract;
    property Id: string read GetId;
    property CreatedAt: TDateTime read FCreatedAt;
  end;

  { A user entity with full feature coverage. }
  TUser = class(TEntity, ISerializable)
  private
    FName: string;
    FEmail: string;
    FLevel: TLogLevel;
    procedure LogAction(const Action: string);
  protected
    FIsActive: Boolean;
  public
    constructor Create(const AId, AName, AEmail: string);
    destructor Destroy; override;
    function Validate: Boolean; override;
    function ToJSON: string;
    function FetchProfile: string;
    property Name: string read FName write FName;
    property Level: TLogLevel read FLevel write FLevel;
  end;

{ Top-level function. }
function PointDistance(const A, B: TPoint): Double;

{ Top-level procedure. }
procedure LogMessage(const Msg: string; Level: TLogLevel);

implementation

{ TEntity }

constructor TEntity.Create(const AId: string);
begin
  FId := AId;
  FCreatedAt := Now;
end;

destructor TEntity.Destroy;
begin
  inherited;
end;

function TEntity.GetId: string;
begin
  Result := FId;
end;

{ TUser }

constructor TUser.Create(const AId, AName, AEmail: string);
begin
  inherited Create(AId);
  FName := AName;
  FEmail := AEmail;
  FLevel := llInfo;
  FIsActive := True;
end;

destructor TUser.Destroy;
begin
  inherited;
end;

function TUser.Validate: Boolean;
begin
  Result := (FName <> '') and (Pos('@', FEmail) > 0);
end;

function TUser.ToJSON: string;
begin
  Result := Format('{"id":"%s","name":"%s"}', [FId, FName]);
end;

function TUser.FetchProfile: string;
begin
  LogAction('fetch_profile');
  WriteLn('Fetched profile for ', FName);
  Result := ToJSON;
end;

procedure TUser.LogAction(const Action: string);
begin
  WriteLn(Format('[%d] %s: %s', [Ord(FLevel), FName, Action]));
end;

{ Top-level routines }

function PointDistance(const A, B: TPoint): Double;
var
  DX, DY: Double;
begin
  DX := A.X - B.X;
  DY := A.Y - B.Y;
  Result := Sqrt(DX * DX + DY * DY);
end;

procedure LogMessage(const Msg: string; Level: TLogLevel);
begin
  WriteLn(Format('[%d] %s', [Ord(Level), Msg]));
end;

end.
